### PR TITLE
planner: fix the bug that wrong collation is used when try fast path for enum or set

### LINKED ISSUE #23053: wrong fast plan for set or enum
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t1(c set("A", "B") collate utf8mb4_general_ci);
desc select  c   from t1  where c = "a";
```

### 2. What did you expect to see? (Required)
``` sql
mysql> desc select  c   from t1  where c = "a";
+-------------------------+---------+-----------+---------------+--------------------------------+
| id                      | estRows | task      | access object | operator info                  |
+-------------------------+---------+-----------+---------------+--------------------------------+
| Selection_5             | 0.80    | root      |               | eq(test.t1.c, "a")             |
| └─TableReader_7         | 1.00    | root      |               | data:TableFullScan_6           |
|   └─TableFullScan_6     | 1.00    | cop[tikv] | table:t1      | keep order:false, stats:pseudo |
+-------------------------+---------+-----------+---------------+--------------------------------+
3 rows in set (0.00 sec)
```
### 3. What did you see instead (Required)
``` sql 
mysql> desc select  c   from t1  where c = "a";
+-------------+---------+------+---------------+---------------+
| id          | estRows | task | access object | operator info |
+-------------+---------+------+---------------+---------------+
| TableDual_2 | 0.00    | root |               | rows:0        |
+-------------+---------+------+---------------+---------------+
1 row in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

4.0, master

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #23053 

Problem Summary:
When comparing datum, we used the wrong collation.
```
		dVal, err := d.ConvertTo(stmtCtx, &col.FieldType)
		if err != nil {
			if terror.ErrorEqual(types.ErrOverflow, err) {
				return append(nvPairs, nameValuePair{colName: colName.Name.Name.L, value: d, param: param}), true
			}
			// Some scenarios cast to int with error, but we may use this value in point get.
			if !terror.ErrorEqual(types.ErrTruncatedWrongVal, err) {
				return nil, false
			}
		}
		cmp, err := d.CompareDatum(stmtCtx, &dval)
```
CompareDatum uses the d's collation, which is wrong. 
### What is changed and how it works?
d.CompareDatum(stmtCtx, &dval) =>
dVal.CompareDatum(stmtCtx, &d)

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- fix the bug that wrong collation is used when try fast path for enum or set

### CODE REVIEW

File: expression/integration_test.go [hunk 1]
```diff
@@ -6280,6 +6280,11 @@ func (s *testIntegrationSerialSuite) TestCollationBasic(c *C) {
 	tk.MustQuery("select a from t_ci where a='A'").Check(testkit.Rows("a"))
 	tk.MustQuery("select a from t_ci where a='a   '").Check(testkit.Rows("a"))
 	tk.MustQuery("select a from t_ci where a='a                    '").Check(testkit.Rows("a"))
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(c set('A', 'B') collate utf8mb4_general_ci);")
+	tk.MustExec("insert into t values('a');")
+	tk.MustQuery("select c from t1  where c = 'a';").Check(testkit.Rows("A"))
```
Comment (Reminiscent) [inline expression/integration_test.go:6287]: Here should be `select c from t where c = 'a';`

**Review by qw4990 — COMMENTED** [review]
Please fix the CI problem:
```
[2021-03-09T11:40:22.369Z] FAIL: integration_test.go:6242: testIntegrationSerialSuite.TestCollationBasic
``` @wjhuang2016

File: expression/integration_test.go (hunk 1 above)
Comment (qw4990) [inline expression/integration_test.go:6287]: Please add more test queries like [this comment](https://github.com/pingcap/tidb/issues/23053#issuecomment-793592738).

**Review by qw4990 — APPROVED** [review]

**Review by Reminiscent — APPROVED** [review]
/LGTM

**Pull request merged**